### PR TITLE
[5.3][Property Wrappers] For now, don't allow wrappedValue and projectedValue to have dynamic Self type.

### PR DESCRIPTION
**Pull request by hborla**
Cherry-picked from https://github.com/apple/swift/pull/31344 and it's follow-up https://github.com/apple/swift/pull/31443

---
Currently, when wrappedValue or projectedValue have dynamic Self type, the synthesized properties when using the wrapper will also get a dynamic Self type, but in a new context. This bug could manifest as really misleading diagnostics, or a crash. For now, we will disallow this, but we can re-visit implementing correct support for this in the future.

Resolves: rdar://problem/53258469

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckAttr.cpp [hunk 1]
```diff
@@ -2862,7 +2862,7 @@ void AttributeChecker::visitCustomAttr(CustomAttr *attr) {
 
   // If the nominal type is a property wrapper type, we can be delegating
   // through a property.
-  if (nominal->getPropertyWrapperTypeInfo()) {
+  if (nominal->getAttrs().hasAttribute<PropertyWrapperAttr>()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckAttr.cpp:2865]: What does this change do?

File: lib/Sema/TypeCheckPropertyWrapper.cpp [hunk 2]
```diff
@@ -406,6 +406,27 @@ PropertyWrapperTypeInfoRequest::evaluate(
     }
   }
 
+  auto diagnoseInvalidDynamicSelf = [&]() -> bool {
+    bool invalidDynamicSelf = false;
+    if (result.projectedValueVar &&
+        result.projectedValueVar->getValueInterfaceType()->is<DynamicSelfType>()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckPropertyWrapper.cpp:412]: I think you want hasDynamicSelfType() here to catch Optional<Self>, etc

File: lib/Sema/TypeCheckPropertyWrapper.cpp [hunk 3]
```diff
@@ -406,6 +406,27 @@ PropertyWrapperTypeInfoRequest::evaluate(
...
+    if (result.valueVar->getValueInterfaceType()->is<DynamicSelfType>()) {
+      result.valueVar->diagnose(
+          diag::property_wrapper_dynamic_self_type, /*projectedValue=*/false);
+      invalidDynamicSelf = true;
+    }
+
+    return invalidDynamicSelf;
+  };
+
+  if (diagnoseInvalidDynamicSelf())
```
Comment (slavapestov) [inline lib/Sema/TypeCheckPropertyWrapper.cpp:427]: It might be more readable without the local function...

File: lib/Sema/TypeCheckAttr.cpp (hunk 1 above)
Comment (hborla) [reply lib/Sema/TypeCheckAttr.cpp:2865]: This change only suppresses the `cannot be used as attribute` diagnostic in the case where the nominal type for the custom attribute type is declared as `@propertyWrapper` but there's some other error, such as the type not having a non-static `wrappedValue` property

File: lib/Sema/TypeCheckPropertyWrapper.cpp (hunk 2 above)
Comment (hborla) [reply lib/Sema/TypeCheckPropertyWrapper.cpp:412]: Oh, yes, thank you! 🤦🏻‍♀️

**Comment by hborla** [comment]
@swift-ci please test

**Review by DougGregor — APPROVED** [review]

**Pull request merged**